parking to HA/Dec frame

wildi opened the pull request:
Still not satisfied with unpark/park of lx200ap driver I studied the process with the mount simulator. 

The current version parks to RA/Dec frame which is rotating for an earth bound observer. The simulator parks now to HA/Dec frame.

I changed the Angle method lst() from private to public. It comes in handy.

knro commented:
So this would be implemented for LX200AP driver as well? i.e. HA/DE parking?

knro commented:
@Chris-Rowland Please share your thoughts on this PR.

Chris-Rowland commented:
The Angle class has conversions to and from different angle units so there is no need for all the multiplying and dividing by 15.
currentRA = (alignment.lst() - Angle(ParkPositionN[AXIS_RA].value * 15.)).Degrees()/15.;
could be
currentRA = (alignment.lst() - Angle(ParkPositionN[AXIS_RA].value, ANGLE_UNITS::HOURS)).Hours();

I don't understand why park would ever be to a Ra Dec position because this position is rotating relative to the terrestrial coordinate system.

One thing, the HA to Ra conversion in telescopesimulator.cpp : relies on the Alignment class being initialised correctly, in particular having the longitude set. Will this have been done?

No problem with Alignment::lst() being public.

Chris-Rowland commented:
Further comments, sorry!
As the park data type is now HA_DEC should the positions set in SetCurrentPark() and SetDefaultPark change?
SetCurrentPark needs to convert the Axis1 position to the current hour angle and SetDefaultPark probably needs to set the Axis1 position to an HA of 6h or -6h.

This would also apply in initParameters is InitPark fails.

Finally (I hope) no other mounts use PARK_HA_DEC.  Quite a lot use RA_DEC and how they manage is anybody's guess but I suspect they are actually parking to a Ha Dec position using a Ra Dec slew to do so.

Chris-Rowland commented:
And...
Park() needs to convert the Axis1 Ha to Ra in order to do the slew.

None of this really matters for the simulator because there's no hardware moving around all over the place and crashing into things but it would be good to get it right.

knro commented:
Right, all other mounts should use HA/DE or AZ/AL for parking. @Chris-Rowland is there any role INDI::Telescope can play in this to make this easier for drivers, or is the current system good as it is?

Chris-Rowland commented:
The park process should cause the mount to go to a known, consistent in the terrestrial frame, position, stop all movement and allow no movement.

In many cases park is managed by the mount and the positions in the driver are irrelevant.  The actual mount park position may not even be available from the mount.

Other mounts don't have park so Indi simulates this by doing a slew and stopping tracking.  This usually needs a Ra/Dec slew but the park position really needs to be as Ha Dec.  Park() needs to convert the park Ha/Dec to the corresponding RaDec. SetCurrentPark should convert the current RaDec to HaDec.

An AltAz mount would park at an AltAz position, even though it may use RaDec to get there, and again the appropriate conversions need to be done.

And some mounts could park to axis positions directly.

This is all so mount dependent that I don't think that inditelescope can or should be involved, not even to the extent is does now where park position data must be made available which in many cases is irrelevant.  But we are where we are, we can't make major changes so it will be best to leave it as it is.

Documenting TelescopeParkData so developers can be aware of what the different type are for may be useful.

knro commented:
Alright, I'm implementing this in each driver separately. Fortunately, it's just a couple of mount drivers that were using RA_DE style parking.

wildi commented:
Hello

I'm currently on vacation but I'll study all the comments
and implement them accordingly.

I thought lst is a public resource :-), thanks, Chris

Kind regards, bye, wildi


On 15/06/2020 10.49, Chris-Rowland wrote:
> The Angle class has conversions to and from different angle units so there is no need for all the multiplying and dividing by 15.
> currentRA = (alignment.lst() - Angle(ParkPositionN[AXIS_RA].value * 15.)).Degrees()/15.;
> could be
> currentRA = (alignment.lst() - Angle(ParkPositionN[AXIS_RA].value, ANGLE_UNITS::HOURS)).Hours();
> 
> I don't understand why park would ever be to a Ra Dec position because this position is rotating relative to the terrestrial coordinate system.
> 
> One thing, the HA to Ra conversion in telescopesimulator.cpp : relies on the Alignment class being initialised correctly, in particular having the longitude set. Will this have been done?
> 
> No problem with Alignment::lst() being public.
> 
>

Chris-Rowland commented:
Is there a need to include scopesim_helper in the CPP file when it's already in the .h file?

239, 245, 247 - Shouldn't the dec be + or - 90 depending on the latitude?

622, same thing except I now see this is by design. Could be worth pointing this out in UpdateProperties().

Park() needs to convert the Axis1 Ha to Ra in order to do the slew.

wildi commented:
I put in Chris' requests into the driver and I think it can be pulled.

wildi commented:
>Is there a need to include scopesim_helper in the CPP file when it's already in the .h file?
Not mine, removed it anyway

>Could be worth pointing this out in UpdateProperties().
Already mentioned in SetDefaultPark(), added hint in UpdateProperties().

>Park() needs to convert the Axis1 Ha to Ra in order to do the slew.
Was already RA, now using Angle() with unit hours, slews to HA=-6., Dec=0. after selecting e.g. Site management, Park Options  Default.

wildi commented:
I have seen that the pull may be occurred before Chris' last comment. I think it is worth to pull the last changes.
